src/pubsub/ua_pubsub_writer.c

```diff
@@ -1355,4 +1367,165 @@ UA_DataSetWriter_generateDataSetMessage(UA_Server *server,
+        }
+
+        if((u64)dataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_UADPDATASETMESSAGECONTENTMASK_TIMESTAMP) {
+            dataSetMessage->header.timestampEnabled = true;
+            dataSetMessage->header.timestamp = UA_DateTime_now();
+        }
+        /* TODO: Picoseconds resolution not supported atm */
+        if((u64)dataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_UADPDATASETMESSAGECONTENTMASK_PICOSECONDS) {
+            dataSetMessage->header.picoSecondsIncluded = false;
+        }
+
+        /* TODO: Statuscode not supported yet */
+        if((u64)dataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_UADPDATASETMESSAGECONTENTMASK_STATUS) {
+            dataSetMessage->header.statusEnabled = true;
+            dataSetMessage->header.status = UA_STATUSCODE_GOOD;
+        }
+    } else if(messageType == UA_TYPES_JSONDATASETWRITERMESSAGEDATATYPE) {
+        if((u64)jsonDataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_JSONDATASETMESSAGECONTENTMASK_METADATAVERSION) {
+            dataSetMessage->header.configVersionMajorVersionEnabled = true;
+            dataSetMessage->header.configVersionMajorVersion = 0;
+
+        }
+        if((u64)jsonDataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_JSONDATASETMESSAGECONTENTMASK_METADATAVERSION) {
+            dataSetMessage->header.configVersionMinorVersionEnabled = true;
+            dataSetMessage->header.configVersionMinorVersion =
+               0;
+        }
+
+        if((u64)jsonDataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_JSONDATASETMESSAGECONTENTMASK_SEQUENCENUMBER) {
+            dataSetMessage->header.dataSetMessageSequenceNrEnabled = true;
+            dataSetMessage->header.dataSetMessageSequenceNr =
+                dataSetWriter->actualDataSetMessageSequenceCount;
+        }
+
+        if((u64)jsonDataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_JSONDATASETMESSAGECONTENTMASK_TIMESTAMP) {
+            dataSetMessage->header.timestampEnabled = true;
+            dataSetMessage->header.timestamp = UA_DateTime_now();
+        }
+
+        /* TODO: Statuscode not supported yet */
+        if((u64)jsonDataSetWriterMessageDataType->dataSetMessageContentMask &
+           (u64)UA_JSONDATASETMESSAGECONTENTMASK_STATUS) {
+            dataSetMessage->header.statusEnabled = true;
+            dataSetMessage->header.status = UA_STATUSCODE_GOOD;
+        }
+    }
+
+    /* Set the sequence count. Automatically rolls over to zero */
+    dataSetWriter->actualDataSetMessageSequenceCount++;
+    /* Prepare DataSetMessageContent */
+    dataSetMessage->header.dataSetMessageValid = true;
+    dataSetMessage->header.dataSetMessageType = UA_DATASETMESSAGE_DATAKEYFRAME;
+    dataSetMessage->data.keyFrameData.fieldCount = 0; // Heartbeat
+
+    return UA_STATUSCODE_GOOD;
+}
+
```

Comment:
I need for some semantically meaningful functions that are called from this function to reduce its cognitive complexity.

Reply:
Generally this function needs refactoring, at least to bring it to the state equal to UA_DataSetWriter_generateDataSetMessage.